File: mobile-app/lib/core/phone_parameters.dart
/// アバターパラメータ定義(docs/protocol.md「1. アバターパラメータ定義」と一致させること)。
enum ParamType { boolean, integer }

class PhoneParameter {
  const PhoneParameter(this.name, this.type, {this.min = 0, this.max = 0});

  final String name;
  final ParamType type;
  final int min;
  final int max;
}

class PhoneParameters {
  PhoneParameters._();

  static const String visible = 'Phone/Visible';
  static const String connected = 'Phone/Connected';
  static const String locked = 'Phone/Locked';
  static const String page = 'Phone/Page';
  static const String pose = 'Phone/Pose';
  static const String battery = 'Phone/Battery';
  static const String callState = 'Phone/CallState';
  static const String mediaState = 'Phone/MediaState';
  static const String notifyType = 'Phone/NotifyType';
  static const String eventToggle = 'Phone/EventToggle';

  static const List<PhoneParameter> all = [
    PhoneParameter(visible, ParamType.boolean),
    PhoneParameter(connected, ParamType.boolean),
    PhoneParameter(locked, ParamType.boolean),
    PhoneParameter(page, ParamType.integer, min: 0, max: 7),
    PhoneParameter(pose, ParamType.integer, min: 0, max: 5),
    PhoneParameter(battery, ParamType.integer, min: 0, max: 10),
    PhoneParameter(callState, ParamType.integer, min: 0, max: 4),
    PhoneParameter(mediaState, ParamType.integer, min: 0, max: 4),
    PhoneParameter(notifyType, ParamType.integer, min: 0, max: 4),
    PhoneParameter(eventToggle, ParamType.boolean),
  ];

  /// UI表示用ラベル(Page 0-7)。
  static const List<String> pageNames = [
    'ロック', 'ホーム', '通知', '通話', 'カメラ', 'メディア', '設定', '接続エラー',
  ];

  /// UI表示用ラベル(Pose 0-5)。
  static const List<String> poseNames = [
    '収納', '右手', '左手', '耳あて', '自撮り', '両手',
  ];

  /// バッテリー残量% → 0-10 の11段階(docs/protocol.md「値の意味」)。
  static int batteryLevelToStep(int percent) {
    if (percent <= 4) return 0;
    final step = ((percent + 5) / 10).floor();
    return step.clamp(0, 10);
  }
}

File: relay-app/src/VrcPhoneRelay.Core/Parameters/PhoneParameters.cs
namespace VrcPhoneRelay.Core.Parameters;

/// <summary>
/// Phone/ 系全パラメータの静的定義表。docs/protocol.md「1. アバターパラメータ定義」が正。
/// </summary>
public static class PhoneParameters
{
    public const string Prefix = "Phone/";

    public const string Visible = "Phone/Visible";
    public const string Connected = "Phone/Connected";
    public const string Locked = "Phone/Locked";
    public const string Page = "Phone/Page";
    public const string Pose = "Phone/Pose";
    public const string Battery = "Phone/Battery";
    public const string CallState = "Phone/CallState";
    public const string MediaState = "Phone/MediaState";
    public const string NotifyType = "Phone/NotifyType";
    public const string EventToggle = "Phone/EventToggle";

    public static readonly IReadOnlyList<ParameterDefinition> All =
    [
        new(Visible, OscValueType.Bool, ParamValue.Bool(false)),
        new(Connected, OscValueType.Bool, ParamValue.Bool(false)),
        new(Locked, OscValueType.Bool, ParamValue.Bool(true)),
        new(Page, OscValueType.Int, ParamValue.Int(0), Min: 0, Max: 7),
        new(Pose, OscValueType.Int, ParamValue.Int(0), Min: 0, Max: 5),
        new(Battery, OscValueType.Int, ParamValue.Int(10), Min: 0, Max: 10),
        new(CallState, OscValueType.Int, ParamValue.Int(0), Min: 0, Max: 4, ResetOnDisconnect: true),
        new(MediaState, OscValueType.Int, ParamValue.Int(0), Min: 0, Max: 4, ResetOnDisconnect: true),
        new(NotifyType, OscValueType.Int, ParamValue.Int(0), Min: 0, Max: 4, ResetOnDisconnect: true),
        new(EventToggle, OscValueType.Bool, ParamValue.Bool(false), RateLimited: true),
    ];

    private static readonly Dictionary<string, ParameterDefinition> ByName =
        All.ToDictionary(d => d.Name, StringComparer.Ordinal);

    /// <summary>アバター対応判定に最低限必要なパラメータ。</summary>
    public static readonly IReadOnlyList<string> RequiredForSupport = [Visible, Page];

    public static bool TryGet(string name, out ParameterDefinition definition) =>
        ByName.TryGetValue(name, out definition!);

    /// <summary>Phone/Connected はスマホからは操作できない(中継アプリのみが書く)。</summary>
    public static bool IsClientWritable(string name) => name != Connected && ByName.ContainsKey(name);
}
